app/models/document_topics.rb

```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+class DocumentTopics
+  include ActiveModel::Model
+
+  attr_accessor :document, :version, :topics
+
+  def self.find_by_document(document)
+    publishing_api = GdsApi.publishing_api_v2
+    links = publishing_api.get_links(document.content_id)
+    topic_content_ids = links.dig("links", "taxons").to_a
+
+    new(
+      document: document,
+      version: links["version"],
+      topics: topic_content_ids.map(&Topic.method(:find)),
```

Comment:
Sorry to be a pain with late comments.

One thing that doesn't sit happy with me and this implementation is that we call this class method on `Topic` and then that creates a new instance of `TopicIndexService` which itself goes off and makes HTTP requests (or loads data from a cache). This seems inefficient when an instance of TopicIndexService could hold this data in memory. 

I'd be suggesting that when we create a DocumentTopics we pass around a TopicIndexService (or something along those lines) so this data can be stored. What do you think?

Reply:
I'm not really sure what the problem is - the TopicIndexService deals with caching, so there's no need to bother other classes with the same issue?

Reply:
Maybe I didn't explain well.

So if we had one instance of TopicIndexService that stored the tree in memory and was passed around, we'd only ever have to do one one look up of the taxonomy tree. 

As this implementation stands each time we create a Topic we need to retrieve the data from cache all pull the request. Which means:

1) In dev - since you have no cache HTTP requests each time
2) with a file system cache - you have to read from file system each time this is used rather than memory
3) with a redis or memcache server - a HTTP request each cache hit

Normally this wouldn't be the sort of thing that'd be a concern as maybe you hit a cache once or twice in a request duration but here we have loops of things doing this.

I'm probably on the verge of premature optimisation but it just sits weird with me and feels more like a candidate for dependency resolution. Does that make sense?

Reply:
Ah, I understand what you mean now, and can see we were operating under different assumptions. In particular, the Rails FileStore does [have an in-memory cache](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/cache/file_store.rb#L15).

But I see what you mean that a Topic could be more conscious of the service it's relying on, but at the expense of passing it around everywhere, which I think is what lead you on to dependency resolution.

My original thought was that the TopicIndexService is acting like a 'bare metal' adapter for a topic, but this has made me think that really I'm treating it more like a TopicIndex model. It's a similar compromise between doing stuff in Ruby vs. SQL.

Without making the code more complicated, I think the best solution to this would be to download the topics index so that it can truly be a model, which makes the overhead of requesting all the data negligible - kind of the same as the original sync/async discussion we had about taxonomy. While the caching is sufficient, I think we're OK for now.

Does that address your concern (at least in principle)?